generator/src/main/java/org/corfudb/generator/Generator.java

```diff
@@ -19,55 +19,196 @@
+    static final int QUEUE_CAPACITY = 100000;
+
+    static final Duration CHUNK_RUNTIME = Duration.ofHours(1);
+    static final Duration CHUNK_TIMEOUT = Duration.ofSeconds(5);
+    static final Duration POLL_TIME = Duration.ofSeconds(3);
+    static final Duration LIVENESS_TIMEOUT = Duration.ofMinutes(15);
+    static final long TIME_TO_WAIT_FOR_RUNTIME_TO_CONNECT = 60000;
+
+
+    public Generator(String endPoint, long cpPeriod, Duration totalRuntime) {
+
+        this.cpPeriod = cpPeriod;
+        this.totalRuntime = totalRuntime;
+
+        rt = new CorfuRuntime(endPoint).connect();
+
+        try {
+            tryToConnectTimeout(TIME_TO_WAIT_FOR_RUNTIME_TO_CONNECT);
+        } catch (SystemUnavailableError e) {
+            System.exit(1);
+        }
+
+        state = new State(numStreams, numKeys, rt);
+
+        scheduler = Executors.newScheduledThreadPool(1);
+
+        workers = newWorkStealingPool(numThreads);
+        operationQueue = new ArrayBlockingQueue<>(QUEUE_CAPACITY);
+        operationQueueLock = false;
+    }
+
+    /**
+     * Try to connect the runtime and throws a SystemUnavailableError if cannot connect
+     * within the timeout.
+     *
+     * @param timeoutInMs
+     * @throws SystemUnavailableError
+     */
+    private void tryToConnectTimeout (long timeoutInMs) throws SystemUnavailableError {
+        try {
+            CompletableFuture.supplyAsync(() -> rt.connect()).get(timeoutInMs, TimeUnit.MILLISECONDS);
+        } catch (Exception e) {
+            Correctness.recordOperation("Liveness, " + "Fail", false);
+            throw new SystemUnavailableError(e.getMessage());
+        }
+    }
+
+    /**
+     * A method to generate operations, and then put the operations into the blocking queue.
+     * The blocking queue serves as a regulator for how many operations exist at the same time.
+     */
+    private void fillOperationQueue() {
+        // Fill the operationQueue to its full capacity
+        List<Operation> operations = state.getOperations().sample(QUEUE_CAPACITY);
 
-        State state = new State(numStreams, numKeys, rt);
+        operationQueueLock = true;
 
+        for (Operation operation : operations) {
+            try {
+                operationQueue.put(operation);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                break;
+            } catch (Exception e) {
+                log.error("operation error", e);
+            }
+        }
+        operationQueueLock = false;
+    }
+
+    /**
+     * Each workers in the worker pool will consume an operation from the blocking queue.
+     * Once the operation de-queued, it is executed.
+     */
+    private void consumeOperationQueue() {
+        // Assign a worker for each thread in the pool
+        for (int i = 0; i < numThreads; i++) {
+            workers.execute(() -> {
+                while (!operationQueueLock && !operationQueue.isEmpty()) {
```

Comment:
There might be a serious problem here: once any of these two checks failed, the thread will finish and exit, but we expect it to block until the fillOperationQueue put more operations into the queue and continue consuming, right? I think that's why we are using a  blocking queue.

Reply:
Fixed.